phases/01-math-foundations/12-tensor-operations/code/notes.py

```
@@ -0,0 +1,29 @@
+    def __init__(self, data, shape=None):
+        if isinstance(data, (list, tuple)):
+            self._data, self._shape = self._flatten_nested(data)
+        elif isinstance(data, np.ndarray):
+            self._data = data.flatten().tolist()
+            self._shape = tuple(data.shape)
+        else:
+            self._data = [data]
+            self._shape = ()
+
+        if shape is not None:
+            total = reduce(lambda a, b: a * b, shape, 1)
```

Ruff (0.16.1)

[error] Undefined name `reduce`

(F821)